pkg/ebpf/c/tracer-ebpf.c

```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
+        }
+    }
+
+    batch->pos++;
+    log_debug("http response ended: code: %d duration: %d(ms)\n", http->response_code, (http->response_last_seen-http->request_started)/(1000*1000));
+}
+
+static __always_inline int http_begin_request(http_transaction_t *http, http_state_t new_state, char *buffer) {
+    // This can happen in the context of HTTP keep-alives;
+    if (http->state == HTTP_RESPONDING) {
+        http_end_response(http);
+    }
+
+    log_debug("http request started\n");
+    http->state = new_state;
+    http->request_started = bpf_ktime_get_ns();
+    http->response_last_seen = 0;
+    http->response_code = 0;
+    __builtin_memcpy(&http->request_fragment, buffer, HTTP_BUFFER_SIZE);
+    return 1;
+}
+
+static __always_inline int http_begin_response(http_transaction_t *http, char *buffer) {
+    // We missed the corresponding request so nothing to do
+    if (!(http->state&HTTP_REQUESTING)) {
+        return 0;
+    }
+
+    // Find digit following space and multiply it by 100 to obtain the response code
+    // HTTP/1.1 200 OK
+    // _________^_____
+    __u16 response_code = 0;
+#pragma unroll
+    for (int i = 0; i < HTTP_BUFFER_SIZE-1; i++) {
+        if (response_code == 0 && buffer[i] == ' ') {
+            response_code = (buffer[i+1]-'0')*100;
+        }
+    }
+
+    if (response_code < 100 || response_code >= 600) {
+        return 0;
+    }
+
+    http->state = HTTP_RESPONDING;
+    http->response_code = response_code;
+    http->response_last_seen = bpf_ktime_get_ns();
+    log_debug("http response started: code: %d\n", http->response_code);
+    return 1;
+}
+
+static __always_inline http_state_t http_read_data(struct __sk_buff* skb, skb_info_t* skb_info, char* p) {
+#pragma unroll
+    for (int i = 0; i < HTTP_BUFFER_SIZE; i++) {
+        if (skb_info->data_off + i <= skb_info->data_end) {
+            p[i] = load_byte(skb, skb_info->data_off + i);
+        } else {
+            // TODO: follow-up on why the verifier rejects the program without this else branch
+            // even if you memset the buffer before
+            p[i] = '\0';
+        }
+    }
+
+    http_state_t packet_type = HTTP_UNKNOWN;
+    if ((p[0] == 'H') && (p[1] == 'T') && (p[2] == 'T') && (p[3] == 'P')) {
+        packet_type = HTTP_RESPONDING;
+    } else if ((p[0] == 'G') && (p[1] == 'E') && (p[2] == 'T')) {
+        packet_type = HTTP_REQUESTING_GET;
+    } else if ((p[0] == 'P') && (p[1] == 'O') && (p[2] == 'S') && (p[3] == 'T')) {
+        packet_type = HTTP_REQUESTING_POST;
+    } else if ((p[0] == 'P') && (p[1] == 'U') && (p[2] == 'T')) {
+        packet_type = HTTP_REQUESTING_PUT;
+    } else if ((p[0] == 'D') && (p[1] == 'E') && (p[2] == 'L') && (p[3] == 'E') && (p[4] == 'T') && (p[5] == 'E')) {
+        packet_type = HTTP_REQUESTING_DELETE;
+    } else if ((p[0] == 'H') && (p[1] == 'E') && (p[2] == 'A') && (p[3] == 'D')) {
+        packet_type = HTTP_REQUESTING_HEAD;
+    }
```

Comment:
Are we planning to support the other common methods? `CONNECT`, `OPTIONS`, and `PATCH`?